### DISCUSSION TITLE
How to set filters and address modes for loaded textures/samplers?

### DISCUSSION THREAD

**Discussion by agausmann** [bevyengine/bevy#1181] [Q&A] [answered]
I've been searching for an answer to this for hours, and as far as I can tell, there is no example or project that shows how to do this, and it's not clear from the bevy code itself. I'm out of ideas at this point.

I started by looking at `SamplerDescriptor`, which has fields for setting `AddressMode`s and `FilterMode`s. A `Texture` has a `SamplerDescriptor`, so I could set it there in theory. But it's not so straightforward to access. Loading a texture the conventional way, using `AssetServer::load`, gives a `Handle<Texture>`, which doesn't dereference directly.

I could pass the handle to `Assets<Texture>::get_mut` to get a `&mut Texture`, but that only works once the texture has been loaded, which probably hasn't happened by the time I'd like to call it (immediately after `load`).

There's no way to figure out when the texture has loaded either. I tried looping on the value of `AssetServer::get_load_state`, but that seems to block progress, being stuck in an infinite `Loading` loop. There's also no obvious `and_then` that I could use to modify the texture immediately after it is loaded and before it is handed off to the registry.

So how then would I do it? Would I need to make my own `AssetLoader` wrapping something like an `ImageTextureLoader`? That doesn't seem like it would work either, I still couldn't get a reference to the created `Texture` object.

**Comment by DJMcNab** [comment, marked as answer]
You can track `Events<AssetEvent<Texture>>`, which will send an event when your texture has loaded.

**Reply by agausmann** [reply]
Ah ha, I thought there might be an event but I didn't look very closely at that. Thank you!
